# Week-1 Assignment

*Welcome to your first assignment for the SimuTech Winter Project 2022! I hope you are excited to implement and test everything you have learned up until now. There is an interesting set of questions for you to refine your acquired skills as you delve into hands-on coding and deepen your understanding of numpy, pandas, and data visualization libraries.*

# Section 0 : Importing Libraries

*Let's begin by importing numpy, pandas and matplotlib.*

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Section 1 : Playing with Python and Numpy

### Q1. Matrix Multiplication

##### (i) Check if matrix multiplication is valid

In [2]:
import numpy as np
def isValid(A,B):
  sA=A.shape
  sB=B.shape
  if (sA[1]==sB[0]):
   return True
  else:
   return False

##### (ii) Using loops (without using numpy)

In [3]:
def matrix_multiply(A,B):
  C = [[0 for _ in range(len(B[0]))] for _ in range(len(A))]
  for i in range(len(A)):
    for j in range(len(B[0])):
      C[i][j]=0
  for i in range(len(A)):
    for j in range(len(B[0])):
      for k in range(len(B)):
        C[i][j]+=A[i][k]*B[k][j]
  return C


##### (iii) Using numpy

In [4]:
def matrix_multiply_2(A,B):
  C=np.matmul(A,B)
  return C

##### (iv) Testing your code

Run the following cell to check if your functions are working properly.

*Expected output:*
[ [102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600] ]

In [5]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [10, 11, 12]
])

B = np.array([
    [13, 14, 15],
    [16, 17, 18],
    [19, 20, 21]
])

if isValid(A,B):
  print(f"Result using loops: {matrix_multiply(A,B)}")
  print(f"Result using numpy: {matrix_multiply_2(A,B)}")
else:
  print(f"Matrix multiplication is not valid")

Result using loops: [[102, 108, 114], [246, 261, 276], [390, 414, 438], [534, 567, 600]]
Result using numpy: [[102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600]]


### Q2. Z-Score Normalisation

Z-score normalization refers to the process of normalizing every value in a dataset such that the mean of all of the values is 0 and the standard deviation is 1.

We use the following formula to perform a z-score normalization on every value in a dataset:

New value = (x – μ) / σ

where:

x: Original value

μ: Mean of data

σ: Standard deviation of data

##### (i) Without using numpy

In [6]:
def mean(x):
  s=0
  for i in x:
    s=s+i
  return s/len(x)


In [7]:
def standard_deviation(x):
  m=mean(x)
  s=0
  for i in x:
    s=s+(i-m)**2
  return (s/len(x))**0.5

In [8]:
def zscore_normalisation(x):
  m=mean(x)
  s=standard_deviation(x)
  for i in range(len(x)):
    x[i]=(x[i]-m)/s
  return x

##### (ii) Using numpy

Numpy has in_built functions for calculating mean and standard deviation

In [9]:
def zscore_normalisation_2(x):
  m=np.mean(x)
  s=np.std(x)
  for i in range(len(x)):
    x[i]=(x[i]-m)/s
  return x

##### (iii) Testing your code

Run the following cell to check if your functions are working properly.

*Expected Output:* [-1.06753267 -0.99745394 -0.99745394 -0.81057732 -0.41346451 -0.06307086
  0.31068237  0.91803138  1.22170588  1.89913361]

In [10]:
x = [4, 7, 7, 15, 32, 47, 63, 89, 102, 131]
print(f"Result without using numpy: {zscore_normalisation(x)}")
print(f"Result using numpy: {zscore_normalisation_2(x)}")

Result without using numpy: [-1.0675326683028088, -0.9974539373420117, -0.9974539373420117, -0.8105773214465528, -0.41346451266870277, -0.06307085786471743, 0.3106823739262003, 0.9180313755864415, 1.2217058764165623, 1.8991336090376005]
Result using numpy: [-1.0675326683028088, -0.9974539373420116, -0.9974539373420116, -0.8105773214465527, -0.41346451266870266, -0.06307085786471735, 0.3106823739262004, 0.9180313755864417, 1.2217058764165623, 1.8991336090376005]


### Q3. Sigmoid fn and its derivative

The sigmoid function is a mathematical function that maps any input value to a value between 0 and 1.

It is defined mathematically as s(x) = 1/(1+e^(-x)).

##### (i) Write a fn to implement sigmoid fn

In [11]:
def sigmoidfn(x):
  from math import e
  for i in range(len(x)):
    for j in range(len(x[0])):
      x[i][j]=1/(1+e**(-x[i][j]))
  return x

##### (ii) Write a fn to implement derivative of sigmoid fn

In [12]:
def derivative(x):
  for i in range(len(x)):
    for j in range(len(x[0])):
      x[i][j]=x[i][j]*(1-x[i][j])
  return x

##### (iii) Test your code

Run the following cell to check if your functions are working properly.

*Expected output:*

x on applying sigmoid activation fn is: [ [0.99987661 0.88079708 0.99330715 0.5        0.5       ]
 [0.99908895 0.99330715 0.5        0.5        0.5       ] ]

x on applying derivative of sigmoid activation fn is: [ [-1.23379350e-04 -1.04993585e-01 -6.64805667e-03 -2.50000000e-01
  -2.50000000e-01]
 [-9.10221180e-04 -6.64805667e-03 -2.50000000e-01 -2.50000000e-01
  -2.50000000e-01] ]

In [13]:
x = np.array([
    [9,2,5,0,0],
    [7,5,0,0,0]
])
print(f"x on applying sigmoid activation fn is: {sigmoidfn(x)}")
print(f"x on applying derivative of sigmoid activation fn is: {derivative(x)}")

x on applying sigmoid activation fn is: [[0 0 0 0 0]
 [0 0 0 0 0]]
x on applying derivative of sigmoid activation fn is: [[0 0 0 0 0]
 [0 0 0 0 0]]


# Section 2: Exploring Pandas

*You have been provided with a dataset which includes information about properties of superheated vapor.*

*The dataset consists of the thermophysical properties: specific volume, specific internal energy, specific enthalpy, specific entropy of superheated vapor.*

*Pressure is in kPa and Temperature in centigrade. In the dataframe 75, 100, 125, etc. are temperatures.*

### Read the csv file


In [15]:
import pandas as pd
data=pd.read_csv('/content/superheated_vapor_properties2.csv')

### Display the shape of data frame


In [16]:
data=pd.DataFrame(data)
data.shape

(543, 37)

### Return an array containing names of all the columns

In [17]:
data.columns

Index(['Pressure', 'Property', 'Liq_Sat', 'Vap_Sat', '75', '100', '125', '150',
       '175', '200', '220', '225', '240', '250', '260', '275', '280', '290',
       '300', '320', '325', '340', '350', '360', '375', '380', '400', '425',
       '450', '475', '500', '525', '550', '575', '600', '625', '650'],
      dtype='object')

### Display the number of null values in each column of the dataframe



In [18]:
null_count=data.isnull().sum()
print(null_count)

Pressure      0
Property      0
Liq_Sat       0
Vap_Sat       0
75          527
100         507
125         487
150         447
175         383
200         319
220         407
225         399
240         407
250         343
260         383
275         339
280         379
290         487
300          59
320         479
325         135
340         476
350          67
360         476
375         203
380         476
400           0
425         204
450           0
475         204
500           0
525         272
550           0
575         340
600           0
625         476
650           0
dtype: int64


### Create a column which contains the Pressure and Property columns, seperated with 'at' (For eg. V at 1, H at 101.325). Using this print the following:
- Enthalpy at 75 kPa and 573 K
- Entropy at 493 K and 250 kPa



In [19]:
data['addition']=data['Property'].astype(str)+' at '+data['Pressure'].astype(str)


### Find out the column with the highest number of missing values

In [20]:
null_count.idxmax()

'75'

### What is the average enthalpy of Sat. Liq. at all different pressures in the dataset?

In [21]:
data['Liq_Sat'].mean()

469.0913154696132

### Separate out the V,U,H,S data from the dataset into V_data, U_data, H_data, S_data

In [22]:
V_data=data[(data['Property']=='V')]
U_data=data[(data['Property']=='U')]
H_data=data[(data['Property']=='H')]
S_data=data[(data['Property']=='S')]

# Section 4 : Conculsion




*Congratulations on reaching this point! I hope you had fun solving your first assignment and have also built confidence in applying these libraries. If you are wondering, we will cover more about z-score normalization in Week 2, and the sigmoid function will be used in Week 3. After completing this assignment, you are now prepared to learn about machine learning techniques and implement your own machine learning models.*